### PR TITLE
[Executor] MVHashMap for resource groups

### PR DESCRIPTION
**Pull request by gelash**
For motivation, refer back to:
https://github.com/aptos-labs/aptos-core/pull/9927

1. This PR isolates some pieces from e2e resource group refactoring, mainly the new multi-version data-structure that is able to return speculative size, as well as commit the resource groups and return pointers to the contents at the committed version. Comes with unit tests and integration w. MVHashMap proptests (proptest only checks add/remove/estimate logic, treating resource group with a single tag as a normal location)
2. Refactor the MVHashMap interfaces in general, and support internal caching of storage-level value (using ShiftedTxnIndex utility in both maps) - this is important as the AggregatorV2 identifier lifting has to happen once.
3. Adjust view interfaces to prepare for the integration w. Block-STM and e2e.

Next PR will integrate the view traits in BlockSTM, add struct to VMChangeSet and a trait to get it, and get the resource groups integrated with the full blockstm proptests based on mock executor (here we can test size/exists/metadata too).
Final PR will connect to the session / data_cache / adapter & enable.

### CODE REVIEW

File: aptos-move/mvhashmap/src/versioned_group_data.rs [hunk 1]
```diff
@@ -0,0 +1,667 @@
...
+                .versioned_map
+                .get(tag)
+                .expect("Versioned map entry for a member tag must exist")
+                .range(ShiftedTxnIndex::zero()..ShiftedTxnIndex::new(txn_idx))
+                .next_back()
+            {
+                Some((idx, entry)) => {
+                    if entry.flag == Flag::Estimate {
+                        Err(MVGroupError::Dependency(
+                            idx.idx().expect("May not depend on storage version"),
```
Comment (igor-aptos) [inline aptos-move/mvhashmap/src/versioned_group_data.rs:233]: shouldn't the comment be "base value cannot be marked as estimate" ?

**Review by igor-aptos — APPROVED** [review]
looks good, mostly refactor. 

VersionedGroupData looks good. some minor comments inline

File: aptos-move/mvhashmap/src/versioned_data.rs [hunk 2]
```diff
@@ -238,30 +230,48 @@ impl<K: Hash + Clone + Debug + Eq, V: TransactionWrite> VersionedData<K, V> {
...
+        let prev_entry = v.versioned_map.insert(
+            ShiftedTxnIndex::zero(),
+            CachePadded::new(Entry::new_write_from(0, data)),
+        );
 
+        // Assert that the previous entry for txn_idx, if present, had lower incarnation.
+        assert!(prev_entry.map_or(true, |entry| -> bool {
+            if let EntryCell::Write(i, v) = &entry.cell {
+                // Assert the length of bytes for efficiency (instead of full equality)
+                *i == 0 && v.bytes_len() == bytes_len
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_data.rs:263]: multiple threads might end up setting the base value because of the race of when they observed Uninitialized

File: aptos-move/mvhashmap/src/versioned_data.rs [hunk 3]
```diff
@@ -238,30 +230,48 @@ impl<K: Hash + Clone + Debug + Eq, V: TransactionWrite> VersionedData<K, V> {
...
+    pub fn provide_base_value(&self, key: K, data: V) {
+        let mut v = self.values.entry(key).or_default();
+        let bytes_len = data.bytes_len();
+        // For base value, incarnation is irrelevant, set to 0.
+        let prev_entry = v.versioned_map.insert(
+            ShiftedTxnIndex::zero(),
+            CachePadded::new(Entry::new_write_from(0, data)),
+        );
 
+        // Assert that the previous entry for txn_idx, if present, had lower incarnation.
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_data.rs:259]: uh braindead copypasta, good catch

File: aptos-move/mvhashmap/src/types.rs [hunk 4]
```diff
@@ -61,3 +80,136 @@ pub enum MVModulesOutput<M, X> {
...
+pub(crate) struct ShiftedTxnIndex {
+    idx: TxnIndex,
+}
+
+impl ShiftedTxnIndex {
+    pub fn new(real_idx: TxnIndex) -> Self {
+        Self { idx: real_idx + 1 }
+    }
+
+    pub(crate) fn idx(&self) -> Result<TxnIndex, StorageVersion> {
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/mvhashmap/src/types.rs:97]: Definitely it needs a custom type, I liked Result because the error is indicative that it is actually a version, just "storage", as opposed to "I was supposed to sometimes have an index / version, seems I don't have it right now". Let's do in a later polishing pass so it doesn't become a distraction, but doing it nicely would be good for sure.

File: aptos-move/mvhashmap/src/versioned_group_data.rs [hunk 5]
```diff
@@ -0,0 +1,667 @@
...
+    /// In summary, existing resources may not be added to a group, and existing members
+    /// may not be removed from a group. However, a newly created resource may become
+    /// a member, and also only members whose corresponding instances exist will be stored
+    /// in a given group at a given time.
+    ///
+    /// We maintain the list of all tags the key has been associated with (either
+    /// in storage, or in any of the writes), and also use it to determine the 'group_size'
+    /// when requested. Group size is computed as the sum of sizes (length of raw bytes)
+    /// of all latest inner values of all member tags, plus serialized length of tags.
+    member_tags: HashSet<T>,
```
Comment (georgemitenkov) [inline aptos-move/mvhashmap/src/versioned_group_data.rs:63]: Why do we need these, isn't it always equal to the key set of versioned_map?

**Review by georgemitenkov — APPROVED** [review]

File: aptos-move/mvhashmap/src/types.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/mvhashmap/src/types.rs:97]: added a todo

File: aptos-move/block-executor/src/executor.rs [hunk 6]
```diff
@@ -207,37 +212,43 @@ where
...
-            match versioned_cache.fetch_data(r.path(), idx_to_validate) {
-                Ok(Versioned(version, _)) => r.validate_version(version),
+            if r.is_speculative_failure() {
+                return false;
+            }
+
+            match versioned_cache.data().fetch_data(r.path(), idx_to_validate) {
+                Ok(Versioned(version, _)) => r.validate_versioned(version),
                 Ok(Resolved(value)) => r.validate_resolved(value),
+                Err(Uninitialized) => {
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:238]: We assume that Uninitialized will lead to an initialization, or in other words, read from storage cannot reach the VM without going through the initializing and reading from the MVHashMap. Let me add a comment about that.

File: aptos-move/block-executor/src/executor.rs [hunk 7]
```diff
@@ -207,37 +212,43 @@ where
...
-                Err(Dependency(_)) | Err(Unresolved(_)) => false,
-                Err(NotFound) => r.validate_storage(),
-                // We successfully validate when read (again) results in a delta application
-                // failure. If the failure is speculative, a later validation will fail due to
-                // a read without this error. However, if the failure is real, passing
-                // validation here allows to avoid infinitely looping and instead panic when
-                // materializing deltas as writes in the final output preparation state. Panic
-                // is also preferable as it allows testing for this scenario.
-                Err(DeltaApplicationFailure) => r.validate_delta_application_failure(),
+                Err(Dependency(_)) | Err(Unresolved(_)) | Err(DeltaApplicationFailure) => false,
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:251]: This is while validating - if we see a dependency, Unresolved or DeltaApplicationFailure, we can just consider that this doesn't match whatever was the original read (that is being validated)

File: aptos-move/mvhashmap/src/versioned_group_data.rs (hunk 5 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_group_data.rs:63]: oh my god, good point :D :D thanks

Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_group_data.rs:63]: I think when I did this, I hadn't combined base and updated values with shiftedindex and forcing initialization, and then I never noticed.

File: aptos-move/mvhashmap/src/lib.rs [hunk 8]
```diff
@@ -33,101 +32,57 @@ mod unit_tests;
...
+            group_data: VersionedGroupData::new(),
             modules: VersionedModules::new(),
         }
     }
 
-    pub fn take(self) -> (VersionedData<K, V>, VersionedModules<K, V, X>) {
-        (self.data, self.modules)
+    // TODO: drop pointers, and take data as well.
+    pub fn take(self) -> VersionedModules<K, V, X> {
+        self.modules
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/mvhashmap/src/lib.rs:61]: I think we don't use take at all, executablestore will need modules, but can add that later, maybe we will also know more whether we need data or group data

File: types/src/write_set.rs [hunk 9]
```diff
@@ -168,22 +168,54 @@ impl WriteOp {
...
+    }
+
     fn as_state_value(&self) -> Option<StateValue> {
         self.bytes().map(|bytes| match self.metadata() {
             None => StateValue::new_legacy(bytes.clone()),
             Some(metadata) => StateValue::new_with_metadata(bytes.clone(), metadata.clone()),
         })
     }
+
+    fn from_state_value(maybe_state_value: Option<StateValue>) -> Self {
```
[the remark this answers is not in the document]
Comment (gelash) [reply types/src/write_set.rs:209]: I added a lengthy comment to explain that this is just for the purposes of converting a statevalue to some representative instance w : W that when applied to storage would have the desired effect, which can be useful e.g. to have sentinel elements in the datastructures (like MVHashMap). And it doesn't guarantee which instance is provided if many have the property.

File: aptos-move/aptos-vm/src/data_cache.rs [hunk 10]
```diff
@@ -664,32 +665,36 @@ mod tests {
...
-        // Release the cache to test contents, and to avoid assert when querying key_1 again.
-        let cache = s.release_resource_group_cache();
-        assert_eq!(cache.len(), 1);
-        assert_some!(cache.get(&key_1));
-
-        assert_ok_eq!(s.resource_exists_in_group(&key_1, &tag_2), false);
-    }
+    // TODO: bring back with proper implementation of resource exists in a group. Storage adapter
+    // needs this because resource group change-set preparation will check existence.
+    // #[test]
```
Comment (zekun000) [inline aptos-move/aptos-vm/src/data_cache.rs:670]: this is not the right way of commenting out a test lol, just add #[ignore]

File: aptos-move/aptos-vm-types/src/resolver.rs [hunk 11]
```diff
@@ -51,9 +51,67 @@ pub trait TResourceView {
     }
 }
 
-pub trait ResourceResolver: TResourceView<Key = StateKey, Layout = MoveTypeLayout> {}
+pub trait TResourceGroupView {
+    type Key;
+    type Tag;
```
Comment (zekun000) [inline aptos-move/aptos-vm-types/src/resolver.rs:56]: I don't like the name Tag here it doesn't convey what it is, it probably should be resource_key and group_key

File: aptos-move/block-executor/src/executor.rs [hunk 12]
```diff
@@ -111,17 +115,17 @@ where
     }
 
     fn execute(
-        version: Version,
+        idx_to_execute: TxnIndex,
+        incarnation: Incarnation,
```
Comment (zekun000) [inline aptos-move/block-executor/src/executor.rs:119]: unpack the version seems unrelated changes?

File: aptos-move/block-executor/src/txn_last_input_output.rs [hunk 13]
```diff
@@ -79,24 +77,17 @@ impl<K: Debug + ModulePath> ReadDescriptor<K> {
...
-
     pub fn from_module(access_path: K) -> Self {
         Self {
             access_path,
             kind: ReadKind::Module,
         }
     }
 
-    pub fn from_delta_application_failure(access_path: K) -> Self {
+    pub fn from_speculative_failure(access_path: K) -> Self {
```
Comment (zekun000) [inline aptos-move/block-executor/src/txn_last_input_output.rs:87]: this also seems unrelated change?

File: aptos-move/block-executor/src/view.rs [hunk 14]
```diff
@@ -43,24 +43,25 @@ pub(crate) enum ReadResult<V> {
...
     // Similar to above, but the value was aggregated and is an integer.
     U128(u128),
-    // Read could not resolve the delta (no base value).
-    Unresolved,
-    // Parallel execution halts.
-    ExecutionHalted,
     // Read did not return anything.
-    None,
+    Uninitialized,
+    // Must half the execution of the calling transaction. This might be because
```
Comment (zekun000) [inline aptos-move/block-executor/src/view.rs:48]: half -> halt?

File: aptos-move/mvhashmap/src/types.rs [hunk 15]
```diff
@@ -61,3 +81,136 @@ pub enum MVModulesOutput<M, X> {
     /// if the latest entry corresponded to the module deletion.
     Module((Arc<M>, HashValue)),
 }
+
+// In order to store base vales at the lowest index, i.e. at index 0, without conflicting
+// with actual transaction index 0, the following struct wraps the index and internally
+// increments it by 1.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub(crate) struct ShiftedTxnIndex {
```
Comment (zekun000) [inline aptos-move/mvhashmap/src/types.rs:89]: why do we need this change here? even we want to do this, isn't it better to just shift in the origin of the index from the txn iterator?

**Review by zekun000 — COMMENTED** [review]
I don't like how this pr is organized, it seems carrying a bunch of unrelated changes in a single commit that makes it not very readable. we should at least split small refactors into its own commit

File: aptos-move/block-executor/src/view.rs (hunk 14 above)
Comment (gelash) [reply aptos-move/block-executor/src/view.rs:48]: resolving in follow-up

File: aptos-move/block-executor/src/executor.rs (hunk 12 above)
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:119]: yes, sorry, this went out of control.

File: aptos-move/aptos-vm-types/src/resolver.rs [hunk 16]
```diff
@@ -51,9 +51,67 @@ pub trait TResourceView {
...
+    /// methods return somewhat consistent values (e.g. size != 0 if exists is true), and
+    /// otherwise return an error as the validation is guaranteed to fail.
+    ///
+    /// The collected size is only guaranteed to correspond to the correct size when executed
+    /// from a quiescent, correct state. The result can be viewed as a branch prediction in
+    /// the parallel execution setting, as a wrong value will be (later) caught by validation.
+    /// Thus, R/W conflicts are avoided, as long as the estimates are correct (e.g. updating
+    /// struct members of a fixed size).
+    fn resource_group_size(&self, _key: &Self::Key) -> anyhow::Result<u64> {
+        unimplemented!("TResourceGroupView not yet implemented");
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/aptos-vm-types/src/resolver.rs:94]: just part of breaking down the bigger changes, I probably should have introduced the API later.
